### PR TITLE
Layout variation for separate org pages

### LINKED ISSUE #91: Org logo in all org related page headers
Add org logo in page header on the following pages
- [x] event
- [x] [campId]
- [x] [surId]
- [x] [campId]/events
- [x] contact

### PR DESCRIPTION
**Pull request by kristofferlarberg**
This PR adds the logo of an organization in `PublicHeader` on separate pages related to an organization. This is done by creating two variations of `OrgLayout` one currently used by pages `o/[orgId]`, `o/[orgId]/events` and `o/[orgId]/campaigns` and the other for the rest.

Skärmavbild 2021-03-19 kl  18 39 271

Fixes #91

### CODE REVIEW

File: src/components/layout/OrgLayout.tsx [hunk 1]
```diff
@@ -13,49 +13,69 @@ import OrgHeader from './OrgHeader';
...
+                    </Item>
+                    <Item key="campaigns" title={ tab3 }>
+                        <Content>{ children }</Content>
+                    </Item>
+                </Tabs>
+            </DefaultLayout>
+        );
+    }
     return (
+        //TODO: Break out into separate meta layout component
```
Comment (richardolsson) [inline src/components/layout/OrgLayout.tsx:67]: I'm not entirely sure what this comment refers to, but if it means what I think it means I agree, and I think it should be done as part of this PR. See other comments.

File: src/components/layout/OrgLayout.tsx [hunk 2]
```diff
@@ -13,49 +13,69 @@ import OrgHeader from './OrgHeader';
...
+const OrgLayout = ({ children, mainPage, orgId } : OrgLayoutProps) : JSX.Element => {
     const orgQuery = useQuery(['org', orgId], getOrg(orgId));
     const router = useRouter();
     const path =  router.pathname.split('/');
     let currentTab = path.pop();
 
-    if (currentTab === '[orgId]') {
-        currentTab = 'home';
-    }
+    if (mainPage) {
```
Comment (richardolsson) [inline src/components/layout/OrgLayout.tsx:26]: Instead of wrapping such a huge portion of the functionality in a conditional, we should use composition. I.e. create a separate component for "main pages", e.g. `OrgHomeLayout` and utilize composition the same way that we do for `DefaultLayout` and `OrgLayout` today.

**Review by richardolsson — CHANGES_REQUESTED** [review]
I think there is a better way to architect this that doesn't require a large component which effectively serves two separate sets of purpose, separated by a boolean flag. Instead, we should use two components, one for a basic org layout (without the cover image et c) and another which contains the cover image et c (and composes the former).

File: src/components/layout/MainOrgLayout.tsx [hunk 3]
```diff
@@ -1,6 +1,6 @@
 //TODO: Enable eslint rule and fix errors
 /* eslint-disable  @typescript-eslint/no-non-null-assertion */
-import { Content } from '@react-spectrum/view';
+import { Content } from '@adobe/react-spectrum';
```
Comment (richardolsson) [inline src/components/layout/MainOrgLayout.tsx:3]: What is this change about? Just so I understand what's going on.

File: src/components/layout/SimpleOrgLayout.tsx [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+import DefaultLayout from './DefaultLayout';
+import getOrg from '../../fetching/getOrg';
+
+interface OrgLayoutProps {
+    children: JSX.Element;
+    orgId: string;
+}
+
+const OrgLayout = ({ children, orgId } : OrgLayoutProps) : JSX.Element => {
```
Comment (richardolsson) [inline src/components/layout/SimpleOrgLayout.tsx:14]: The name here should be `SimpleOrgLayout` to reflect the public name of the component (i.e. the one that an importer will know).

**Review by richardolsson — CHANGES_REQUESTED** [review]
I think this turned out much better! There is clearer separation of concerns, all the files are smaller and easier to reason about, and looking at the diff stat it actually meant less changes (and less overwhelming rewrites):

image

I've found some minor issues though:

Both the `SimpleOrgLayout` and `MainOrgLayout` components are not named correctly (but their respective module is).

This is also not exactly how I imagined the structure would be. Since the `SimpleOrgLayout` already takes care of showing the org logo in the header, I did not expect that `MainOrgLayout` would have to do that as well. My idea was to wrap `MainOrgLayout` in `SimpleOrgLayout` the same way that `DefaultLayout` is wrapped in `SimpleOrgLayout`.

Did you consider that and find any issues with it? Just because it was my initial idea doesn't mean it's a good solution. I just want to hear your reasoning.

File: src/components/layout/SimpleOrgLayout.tsx (hunk 4 above)
Comment (kristofferlarberg) [reply src/components/layout/SimpleOrgLayout.tsx:14]: Sorry about this, was a bit too quick in the process of renaming them.

File: src/components/layout/MainOrgLayout.tsx (hunk 3 above)
Comment (kristofferlarberg) [reply src/components/layout/MainOrgLayout.tsx:3]: For some reason the `Content` was imported from `@react-spectrum/view`. `view is a specific and different component from `Content`. I've imported spectrum components from `@adobe/react-spectrum` unless it's necessary with a specific path as with `tabs` component.

**Comment by kristofferlarberg** [comment]
> This is also not exactly how I imagined the structure would be. Since the `SimpleOrgLayout` already takes care of showing the org logo in the header, I did not expect that `MainOrgLayout` would have to do that as well. My idea was to wrap `MainOrgLayout` in `SimpleOrgLayout` the same way that `DefaultLayout` is wrapped in `SimpleOrgLayout`.

Yes, the reason why I did the conditional in `OrgLayout` was so that we would not need to use `DefaultLayout` in several instances, so this makes sense. But just so I understand you correctly, would this mean I remove `DefaultLayout` from `MainOrgLayout` and replace it with `SimpleOrgLayout`?

Also, if we do it like this maybe `DefaultOrgLayout`is a better name for `SimpleOrgLayout`?

**Comment by richardolsson** [comment]
> Yes, the reason why I did the conditional in `OrgLayout` was so that we would not need to use `DefaultLayout` in several instances, so this makes sense. But just so I understand you correctly, would this mean I remove `DefaultLayout` from `MainOrgLayout` and replace it with `SimpleOrgLayout`?

There's nothing inherently negative about using a component in many places. There is merit to trying to reuse things by composing them, but since there is not a lot of actual logic in the `SimpleOrgLayout`, and some of it might need to exist anyway in `MainOrgLayout`, it's not obvious that refactoring it this way will make it better. It's worth considering, but we might just end up deciding to stick with the way you've currently done it.

> Also, if we do it like this maybe `DefaultOrgLayout`is a better name for `SimpleOrgLayout`?

Good point!

Basically, here is a schematic (pseudy-code) of how I imagined it:

```
// Default layout
DefaultLayout
┗ PublicHeader
┗ Content (the page)

// DefaultOrgLayout
DefaultOrgLayout
┗ DefaultLayout
  ┗ (inferred) PublicHeader
  ┗ (inferred) Content (the page)

// MainOrgLayout
MainOrgLayout
┗ DefaultOrgLayout
  ┗ (inferred) DefaultLayout
    ┗ (inferred) PublicHeader
    ┗ (inferred) Content (the page and the tabs, passed in as children from MainOrgLayout)
```

File: src/components/layout/MainOrgLayout.tsx [hunk 5]
```diff
@@ -40,24 +40,26 @@ const OrgLayout = ({ children, orgId } : OrgLayoutProps) : JSX.Element => {
...
-                <Item key="events" title={ tab2 }>
-                    <Content>{ children }</Content>
-                </Item>
-                <Item key="campaigns" title={ tab3 }>
-                    <Content>{ children }</Content>
-                </Item>
-            </Tabs>
-        </DefaultLayout>
+        <DefaultOrgLayout orgId={ orgId }>
+            <>
```
Comment (richardolsson) [inline src/components/layout/MainOrgLayout.tsx:44]: Is this fragment (`<>`) really necessary?

**Review by richardolsson — COMMENTED** [review]
Awesome! This is looking really good.

Only one tiny comment on this. What do you think?

File: src/components/layout/MainOrgLayout.tsx (hunk 5 above)
Comment (kristofferlarberg) [reply src/components/layout/MainOrgLayout.tsx:44]: I added this because only one child was expected. Maybe there's a better way?

Comment (richardolsson) [reply src/components/layout/MainOrgLayout.tsx:44]: Why would you say that only one child was expected? Was there a warning or something?

It's entirely possible and normal in React to have multiple children of an element. In fact, by moving the children into a fragment (`<>` or `React.Fragment`) you still have multiple children, except that they're children of another element (the fragment).

If our layout components are complaining about adding multiple children, we need to change their expectations.

Comment (kristofferlarberg) [reply src/components/layout/MainOrgLayout.tsx:44]: Okay! Yes, the `DefaultOrgLayout` component expected one but got several children.

**Review by richardolsson — APPROVED** [review]
Mission accomplished
